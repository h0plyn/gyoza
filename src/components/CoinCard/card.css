.card-container {
  display: flex;
  flex-direction: row;
  background-color: var(--card-bg);
  margin-bottom: 1rem;
  height: 5rem;
  width: 50vw;
  justify-content: space-around;
  align-items: center;
  transition: transform 0.25s cubic-bezier(0.17, 0.67, 0.83, 0.67) 0s;
}

.content-box {
  /* border: 1px solid red; */
  flex: 1;
  justify-content: center;
  display: flex;
  align-items: center;
  width: 100%;
  height: 100%;
}

.card-container:hover {
  transform: scale(1.03);
  box-shadow: 0 0 40px rgb(192 219 255 / 28%), 0 0 22px var(--highlight);
}

.coin-name {
  display: flex;
  align-items: center;
  justify-content: flex-start;
  gap: 0.5rem;
  padding-left: 2rem;
  font-size: 1.3rem;
  flex: 2;
}

.mkt-cap {
  flex: 2;
}

.card-price {
  display: flex;
  justify-content: flex-start;
}

.up {
  color: seagreen;
}

.down {
  color: lightcoral;
}
