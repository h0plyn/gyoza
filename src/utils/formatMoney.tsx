export function formatMoney(amount: number) {
  if (amount < 0.01) {
    return `$${amount}`;
  } else {
    return new Intl.NumberFormat('en-US', {
      style: 'currency',
      currency: 'USD',
    }).format(amount);
  }
}
