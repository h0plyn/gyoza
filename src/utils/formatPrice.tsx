export function formatMoney(amount: number) {
  const formattedAmount = new Intl.NumberFormat('en-US', {
    style: 'currency',
    currency: 'USD',
  }).format(amount);

  return formattedAmount;
}
