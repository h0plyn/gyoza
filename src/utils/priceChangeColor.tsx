import { Asset } from '../types';
export const priceChangeColor = (coin: Asset): boolean =>
  coin.price_change_percentage_24h.toString().includes('-');
