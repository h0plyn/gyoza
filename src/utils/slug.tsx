import { Asset } from '../types';

export const slug = (coin: Asset): string =>
  coin.id.replace(/\s+/g, '').toLowerCase();
