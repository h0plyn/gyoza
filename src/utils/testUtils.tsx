export const fakeCoin = () => ({
  id: 'ethereum',
  symbol: 'eth',
  name: 'Ethereum',
  image:
    'https://assets.coingecko.com/coins/images/279/large/ethereum.png?1595348880',
  current_price: 2012.98,
  market_cap: 235151119166,
  market_cap_rank: 2,
  fully_diluted_valuation: null,
  total_volume: 23523122647,
  high_24h: 2019.8,
  low_24h: 1888.62,
  price_change_24h: 32.71,
  price_change_percentage_24h: 1.65181,
  market_cap_change_24h: 1926717111,
  market_cap_change_percentage_24h: 0.82612,
  circulating_supply: 116423057.249,
  total_supply: null,
  max_supply: null,
  ath: 4356.99,
  ath_change_percentage: -53.69551,
  ath_date: '2021-05-12T14:41:48.623Z',
  atl: 0.432979,
  atl_change_percentage: 465853.98156,
  atl_date: '2015-10-20T00:00:00.000Z',
  roi: {
    times: 77.02594436717463,
    currency: 'btc',
    percentage: 7702.5944367174625,
  },
  last_updated: '2021-06-24T15:46:48.472Z',
});
