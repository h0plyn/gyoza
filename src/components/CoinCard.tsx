import React, { FC } from 'react';
import { Asset } from '../types';
import { Link } from 'react-router-dom';
import '../styles/card.css';

const CoinCard: FC<{
  coin: Asset;
  setCurrentCoin(coin: Asset): void;
  idx: number;
}> = (props: {
  coin: Asset;
  setCurrentCoin(coin: Asset): void;
  idx: number;
}) => {
  const { coin, idx, setCurrentCoin } = props;
  const slug = coin.id.replace(/\s+/g, '').toLowerCase();
  const mkt_cap_pos = idx + 1;
  const changeColor: boolean = coin.price_change_percentage_24h
    .toString()
    .includes('-');

  return (
    <div className="card-container" key={coin.name}>
      <div className="content-box">{mkt_cap_pos}</div>
      <div
        onClick={() => setCurrentCoin(coin)}
        className="coin-name content-box"
      >
        <img
          src={coin.image}
          alt={coin.name}
          style={{ maxWidth: 20, maxHeight: 20 }}
        />
        <Link to={`${slug}`}>{coin.name}</Link>
      </div>
      <div className="content-box flex card-price">
        ${coin.current_price.toLocaleString()}
      </div>
      <div className={`${changeColor ? 'down' : 'up'} content-box`}>
        {coin.price_change_percentage_24h.toFixed(1)}%
      </div>
      <div className="content-box mkt-cap">
        ${coin.market_cap.toLocaleString()}
      </div>
    </div>
  );
};

export default CoinCard;
