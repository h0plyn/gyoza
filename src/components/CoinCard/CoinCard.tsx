import { FC } from 'react';
import { Link } from 'react-router-dom';
import { CoinCard as CardProps } from '../../types';
import { formatMoney } from '../../utils/formatPrice';
import './card.css';

const CoinCard: FC<CardProps> = (props: CardProps) => {
  const { coin, setCurrentCoin } = props;
  const slug = coin.id.replace(/\s+/g, '').toLowerCase();
  const priceChangeColor: boolean = coin.price_change_percentage_24h
    .toString()
    .includes('-');

  return (
    <div className="card-container" key={coin.name}>
      <div className="content-box" data-testid="coin-rank">
        {coin.market_cap_rank}
      </div>
      <div
        data-testid="coin-name"
        onClick={() => setCurrentCoin(coin)}
        className="coin-name content-box"
      >
        <img
          data-testid="coin-image"
          src={coin.image}
          alt={coin.name}
          style={{ maxWidth: 20, maxHeight: 20 }}
        />
        <Link to={`${slug}`}>{coin.name}</Link>
      </div>
      <div className="content-box flex card-price" data-testid="coin-price">
        {formatMoney(coin.current_price)}
      </div>
      <div className={`${priceChangeColor ? 'down' : 'up'} content-box`}>
        {coin.price_change_percentage_24h.toFixed(1)}%
      </div>
      <div className="content-box mkt-cap" data-testid="coin-market-cap">
        ${coin.market_cap.toLocaleString('en-US')}
      </div>
    </div>
  );
};

export default CoinCard;
