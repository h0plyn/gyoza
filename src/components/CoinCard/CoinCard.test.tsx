import { render, cleanup, fireEvent, waitFor } from '@testing-library/react';
import { BrowserRouter as Router } from 'react-router-dom';
import renderer from 'react-test-renderer';
import CoinCard from './CoinCard';
import { fakeCoin } from '../../utils/testUtils';

describe('<CoinCard />', () => {
  afterEach(cleanup);

  const mockCoin = fakeCoin();

  const setup = () => {
    const setCurrentCoin = jest.fn();
    const utils = render(
      <Router>
        <CoinCard coin={mockCoin} setCurrentCoin={setCurrentCoin} />
      </Router>
    );
    const rank = utils.getByTestId('coin-rank').textContent;
    const name = utils.getByTestId('coin-name');
    const image = utils.getByTestId('coin-image');
    const price = utils.getByTestId('coin-price').textContent;
    const marketCap = utils.getByTestId('coin-market-cap').textContent;
    return { setCurrentCoin, rank, name, image, price, marketCap, ...utils };
  };

  it('should render the CoinCard component without crashing', () => {
    setup();
  });

  it('should match the CoinCard snapshot', () => {
    const tree = renderer.create(
      <Router>
        <CoinCard coin={mockCoin} />
      </Router>
    );
    expect(tree).toMatchSnapshot();
  });

  it('should contain information on the current coin', () => {
    const { rank, name, image, price, marketCap } = setup();
    expect(rank).toBe('2');
    expect(name.textContent).toBe('Ethereum');
    expect(image).toHaveAttribute(
      'src',
      'https://assets.coingecko.com/coins/images/279/large/ethereum.png?1595348880'
    );
    expect(image).toHaveAttribute('alt', 'Ethereum');
    expect(price).toBe('$2,012.98');
    expect(marketCap).toBe('$235,151,119,166');
  });

  it('should set the current coin when the name is clicked', async () => {
    const { setCurrentCoin, name } = setup();
    expect(setCurrentCoin).not.toBeCalled();
    fireEvent.click(name);
    expect(setCurrentCoin).toBeCalledTimes(1);
  });
});
