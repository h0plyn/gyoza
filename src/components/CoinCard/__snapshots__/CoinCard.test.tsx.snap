// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CoinCard /> should match the CoinCard snapshot 1`] = `
<div
  className="card-container"
>
  <div
    className="content-box"
    data-testid="coin-rank"
  >
    2
  </div>
  <div
    className="coin-name content-box"
    data-testid="coin-name"
    onClick={[Function]}
  >
    <img
      alt="Ethereum"
      data-testid="coin-image"
      src="https://assets.coingecko.com/coins/images/279/large/ethereum.png?1595348880"
      style={
        Object {
          "maxHeight": 20,
          "maxWidth": 20,
        }
      }
    />
    <a
      href="/ethereum"
      onClick={[Function]}
    >
      Ethereum
    </a>
  </div>
  <div
    className="content-box flex card-price"
    data-testid="coin-price"
  >
    $2,012.98
  </div>
  <div
    className="up content-box"
  >
    1.7
    %
  </div>
  <div
    className="content-box mkt-cap"
    data-testid="coin-market-cap"
  >
    $
    235,151,119,166
  </div>
</div>
`;
